File: lib/src/geiger_api.dart
library geiger_api;

import 'package:geiger_localstorage/geiger_localstorage.dart';
import 'package:logging/logging.dart';

import '../geiger_api.dart';
import 'plugin/menu_registrar.dart';
import 'plugin/plugin_registrar.dart';

/// The API provided by all communicator interfaces.
abstract class GeigerApi implements PluginRegistrar, MenuRegistrar {
  static const String masterId = '__MASTERPLUGIN__';

  // Is writable to allow examples to specify another master.
  static String masterExecutor = 'FIXME';
  static final Logger logger = Logger("GeigerApi");

  /// Identifier of plugin.
  String get id;

  /// Data sharing [Declaration].
  Declaration get declaration;

  /// [StorageController] to access the master storage.
  StorageController get storage;

  /// Initialize asynchronous parts of the plugin.
  ///
  /// Must be called once after contruction.
  Future<void> initialize();

  /// Retrieve the [StorageController] to access the master storage.
  ///
  /// Throws [StorageException] in case allocation of the storage backend fails.
  @Deprecated('Use [storage]')
  StorageController getStorage() {
    return storage;
  }

  /// Get the [PluginInformation] of all registered plugins.
  ///
  /// Only the plugins which ids start with [startId] are
  /// returned if [startId] is specified.
  /// For security reasons all [PluginInformation.secret]s are empty.
  /// TODO: should only be available on master?
  Future<List<PluginInformation>> getRegisteredPlugins([String? startId]);

  /// Register the [listener] for specific [events] locally.
  ///
  /// Use [MessageType.allEvents] to register to all event types.
  void registerListener(List<MessageType> events, PluginListener listener);

  /// Remove the [listener] from specific [events] locally.
  ///
  /// Set [events] to `null` to remove the [listener] from all events.
  void deregisterListener(List<MessageType>? events, PluginListener listener);

  /// Send a [message] to a another plugin with the id [pluginId].
  ///
  /// If [pluginId] is not specified, [message.targetId] is used.
  /// Non-master plugins can only send messages to the master.
  Future<void> sendMessage(Message message, [String? pluginId]);

  /// Notify plugin about a [MenuItem] with a specific [url] being pressed.
  Future<void> menuPressed(GeigerUrl url);

  /// Get a list of all registered [MenuItem]s.
  ///
  /// Can only be called on the master.
  List<MenuItem> getMenuList();

  /// Notify all plugins that the scan button was pressed.
  ///
  /// Can only be called on the master.
  Future<void> scanButtonPressed();

  /// Reset the GeigerApi by removing all registered plugins and [MenuItem]s.
  Future<void> zapState();

  /// Release all resources.
  Future<void> close();
}

File: java/GeigerApiPluginConnector/src/main/java/eu/cybergeiger/api/GeigerApi.java
package eu.cybergeiger.api;

import eu.cybergeiger.storage.StorageController;
import eu.cybergeiger.storage.StorageException;
import eu.cybergeiger.api.message.GeigerUrl;
import eu.cybergeiger.api.message.Message;
import eu.cybergeiger.api.message.MessageType;
import eu.cybergeiger.api.plugin.*;

import java.io.IOException;
import java.util.List;
import java.util.logging.Logger;

/**
 * <p>The API provided by all communicator interfaces.</p>
 */
public interface GeigerApi extends PluginRegistrar, MenuRegistrar {
  String MASTER_ID = "__MASTERPLUGIN__";
  String MASTER_EXECUTOR = "";

  Logger logger = Logger.getLogger("GeigerApi");

  String getId();

  Declaration getDeclaration();

  /**
   * <p>Obtain controller to access the storage.</p>
   *
   * @return a generic controller providing access to the local storage
   * @throws StorageException in case allocation of storage backed fails
   */
  StorageController getStorage() throws StorageException;

  /**
   * <p>Register an event listener for specific events on the Master.</p>
   *
   * @param events   list of events for which messages should be received. Use MessageType.
   *                 ALL_EVENTS to register for all messages.
   * @param listener the listener to be registered
   */
  void registerListener(MessageType[] events, PluginListener listener);

  /**
   * <p>Remove a listener waiting for Events.</p>
   *
   * @param events   the events affected or null if the listener should be removed from all events
   * @param listener the listener to be removed
   */
  void deregisterListener(MessageType[] events, PluginListener listener);

  /**
   * <p>Sends a custom, plugin-specific message to a peer plugin.</p>
   *
   * <p>Mainly used for internal purposes. Plugins may only send messages to the toolbox core.</p>
   *
   * @param msg the message to be sent
   */
  void sendMessage(Message msg) throws IOException;

  /**
   * <p>Reset the GeigerApi by removing all registered plugins and MenuItems.</p>
   */
  void zapState();

  /**
   * Release all resources.
   */
  void close();
}
